cleverhans/evaluation.py

```diff
@@ -342,4 +379,61 @@ def __call__(self):
     return (x_batch, y_batch), (correct,)
 
 
+class _CorrectAndProbFactory(object):
+  """
+  A factory for an expression for the following tuple per (optionally
+  adversarial) example:
+    - bool per indicating whether each the example was classified correctly
+    - probability assigned to that prediction
+  """
+
+  def __init__(self, model, attack=None, attack_params=None):
+    if attack_params is None:
+      attack_params = {}
+    self.model = model
+    self.attack = attack
+    self.attack_params = attack_params
+    hashable_attack_params = tuple((key, attack_params[key]) for key
+                                   in sorted(attack_params.keys()))
+    self.properties_to_hash = (model, attack, hashable_attack_params)
+
+  def __hash__(self):
+    # Make factory hashable so that no two factories for the
+    # same model will be used to build redundant tf graphs
+    return self.properties_to_hash.__hash__()
+
+  def __eq__(self, other):
+    # Make factory hashable so that no two factories for the
```

Comment:
This comment seems to be copied over from `__hash__`?

Reply:
It is the same comment but it's still valid. An object needs to implement both `__hash__` and `__eq__` to be hashable. The `__hash__` function determines which bucket it lands in and the `__eq__` function makes it possible to tell it apart from other objects in the same bucket.